### ISSUE TITLE
Query size of texture

### ISSUE THREAD

**Issue by dj2** [gpuweb/gpuweb#1107]
In order to implement robust buffer access I need to know the bounds of a given texture to clamp the indices. 

## Proposal

Add the following texture methods:

```
u32 textureSize(texture_ro_1d)
u32 textureSize(texture_wo_1d)
vec2<u32> textureSize(texture_ro_1d_array)
vec2<u32> textureSize(texture_wo_1d_array)
vec2<u32> textureSize(texture_ro_2d)
vec2<u32> textureSize(texture_wo_2d)
vec3<u32> textureSize(texture_ro_2d_array)
vec3<u32> textureSize(texture_wo_2d_array)
vec3<u32> textureSize(texture_ro_3d)
vec3<u32> textureSize(texture_wo_3d)
vec2<u32> textureSize(texture_multisampled_2d)
   OpImageQuerySize

u32 textureSize(texture_sampled_1d, i32 level_of_detail)
vec2<u32> textureSize(texture_sampled_1d_array, i32 level_of_detail)
vec2<u32> textureSize(texture_sampled_2d, i32 level_of_detail)
vec3<u32> textureSize(texture_sampled_2d_array, i32 level_of_detail)
vec3<u32> textureSize(texture_sampled_3d, i32 level_of_detail)
  OpImageQuerySizeLOD

u32 textureMipLevels(texture_ro_1d)
u32 textureMipLevels(texture_wo_1d)
u32 textureMipLevels(texture_ro_1d_array)
u32 textureMipLevels(texture_wo_1d_array)
u32 textureMipLevels(texture_ro_2d)
u32 textureMipLevels(texture_wo_2d)
u32 textureMipLevels(texture_ro_2d_array)
u32 textureMipLevels(texture_wo_2d_array)
u32 textureMipLevels(texture_ro_3d)
u32 textureMipLevels(texture_wo_3d)
  OpImageQueryLevels
```

### Size Method
Size translates to `get_width`, `get_height` and `get_depth`  in Metal Shading Language

Size translates to a `GetDimensions` call in HLSL

### Mip Method
The mip level translates to `get_num_mip_levels` in MSL

The mip level translates to `GetDimensions` in HLSL

**Comment by kvark** [comment]
I agree we need that! A few questions:
  1. Why are we returning `f32` and not `u32`?
  2. Why are we not reporting the array sizes?
  3. Why are we not having it on `texture_multisampled_2d`?
  4. Why are we not having it on `texture_rw_2d`?
  5. Can we also query the mipmap counts (i.e. `get_num_mip_levels()` in MSL)?

**Comment by dj2** [comment]
1- done
2- I don't understand the question
3- done
4- done
5- done (did I get the right types for it?)

**Comment by kvark** [comment]
2. you have `u32 textureSize(texture_ro_1d_array)`. Why is it returning a single value? I'd expect it to return 2 values: the size and the array layer count.
5 - We don't need `u32 textureMipLevels(texture_multisampled_2d)`, it's guaranteed to be non-mipmapped

**Comment by dj2** [comment]
2- done. I also changed the 2d_array to a vec3 to get the array layer count as well.

5- done.

**Comment by dj2** [comment]
Related to #573

**Comment by ben-clayton** [comment]
[I've started a document](https://github.com/ben-clayton/misc/blob/main/shader-texture-functions.md) that lists each of the texture functions by texture type for MSL and HLSL (SPIR-V TODO).

**Comment by jdashg** [comment]
"Should we always return vec3 for textureSize?"
Consensus: Probably not

**Comment by litherum** [comment]
@ben-clayton I did a similar analysis a year or so ago: [Shading Language Standard Libraries.numbers.zip](https://github.com/gpuweb/gpuweb/files/5394989/Shading.Language.Standard.Libraries.numbers.zip)

**Comment by ben-clayton** [comment]
> @ben-clayton I did a similar analysis a year or so ago: [Shading Language Standard Libraries.numbers.zip](https://github.com/gpuweb/gpuweb/files/5394989/Shading.Language.Standard.Libraries.numbers.zip)

Thank you @litherum - I had apparently already downloaded this (probably from [#667](https://github.com/gpuweb/gpuweb/issues/667#issuecomment-608976835)), but my non-mac machine didn't know what to make of most the files in the zip, and I only saw some pictures. I now know "Numbers" is a thing, and opened it on my mac and can see the tables 😃 .

It looks like you have pretty much what I've got (both are currently missing the SPIR-V texture functions). I'll do a quick pass tomorrow of your table against mine, and make sure the HLSL and MSL columns agree. I'll then try filling in my SPIR-V table, and we can decide what to do with this info.

**Comment by dj2** [comment]
This was discussed at the [2020-10-13](https://docs.google.com/document/d/16vNf_rmJ566SoI-iOGGr-UqTGZ32SHKW2_qizoB4HxU/edit#heading=h.hp3f2zbslxr9) WGSL meeting. The consensus was to turn this into a PR with the idea that these may get expanded based on the work @ben-clayton is currently doing.

**Comment by ben-clayton** [comment]
[Document updated with SPIR-V instructions](https://github.com/ben-clayton/misc/blob/main/shader-texture-functions.md).
Will now compare against @litherum 's.

**Comment by kvark** [comment]
Fixed by #1337 and #1348
